File: algorithms/python3/403.py
class Solution:
    def canCross(self, stones):
        """
        :type stones: List[int]
        :rtype: bool
        """
        if stones[1] - stones[0] != 1:
            return False
        m = {stones[1]: set([1])}
        for i in range(1, len(stones)):
            stone = stones[i]
            if stone not in m:
                continue
            for step in m[stone]:
                for nextS in [step-1, step, step+1]:
                    if nextS <= 0:
                        continue
                    if stone + nextS not in m:
                        m[stone + nextS] = set()
                    m[stone + nextS].add(nextS)
        return stones[-1] in m

File: algorithms/rust/403.rs
use std::collections::{HashSet, VecDeque};

impl Solution {
    #[allow(dead_code)]
    pub fn can_cross(stones: Vec<i32>) -> bool {
        let stones: Vec<_> = stones.into_iter().map(|x| x as usize).collect();

        if stones[1] != 1 {
            return false;
        }

        let mut paths = VecDeque::new();
        paths.push_back((stones[1], 1_usize));
        let mut visited = HashSet::new();
        visited.insert((stones[1], 1_usize));

        while let Some(path) = paths.pop_front() {
            let (current, step) = path;

            if Some(&current) == stones.last() {
                return true;
            }

            let (min, max) = (1.max(step - 1), step + 1);
            (min..=max)
                .filter(|&x| stones.binary_search(&(current + x)).is_ok())
                .for_each(|x| {
                    let path = (current + x, x);
                    if !visited.contains(&path) {
                        visited.insert(path);
                        paths.push_back(path);
                    }
                })
        }

        false
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
